jetstream/engine/aot_utils.py

```diff
@@ -0,0 +1,238 @@
+      if bucket <= prefill_engine.max_prefill_length
+  ]
+  prefill_engine.prefill_buckets = prefill_buckets
+  if prefill_engine.max_prefill_length not in prefill_buckets:
+    prefill_buckets.append(prefill_engine.max_prefill_length)
+
+  def compile_prefill(length):
+    metadata = prefill_engine.get_tokenizer()
+    vocab = token_utils.load_vocab(metadata.path, metadata.extra_ids)
+    padded_tokens, true_length = token_utils.tokenize_and_pad(
+        "Example text, often referred to as lorem ipsum, is placeholder content used by designers and developers in the layout of documents and websites. It's a scrambled Latin passage that mimics the rhythm and flow of real text, allowing for accurate visualization of fonts, spacing, and formatting. This nonsensical text helps maintain focus on the visual aspects without distraction from actual content. Lorem ipsum has become a standard in the industry, appearing in countless projects as a temporary stand-in before the final text is incorporated.",  # pylint: disable=line-too-long
+        vocab=vocab,
+        max_prefill_length=length,
+    )
+
+    lowered = jax.jit(
+        prefill_engine.prefill,
+        out_shardings=prefill_engine.get_prefix_destination_sharding(),
+    ).lower(
+        params=prefill_params,
+        padded_tokens=padded_tokens,
+        true_length=true_length,
+    )
+    logging.info(
+        "---------Prefill engine %d lowered for prefill length %d.---------",
+        prefill_idx,
+        length,
+    )
+    compiled = lowered.compile()
+    logging.info(
+        "---------Prefill engine %d compiled for prefill length %d.---------",
+        prefill_idx,
+        length,
+    )
+    prefill_compiled[length] = compiled
+
+  logging.info("---------Prefill compilation %d begun.---------", prefill_idx)
+
+  prefill_compiled = {}
+  with concurrent.futures.ThreadPoolExecutor(
+      max_workers=len(prefill_buckets)
+  ) as executor:
+    _ = executor.map(compile_prefill, prefill_buckets)
+
+  prefill_engine.prefill_compiled = prefill_compiled
+
+  logging.info(
+      "---------Prefill compilation %d complete.---------", prefill_idx
+  )
+
+  return prefill_compiled
+
+
+def initialize_insert_generate_jit_cache(
+    *,
+    generate_engine: engine_api.Engine,
+    generate_params: Any,
+    generate_idx: int,
+):
+  """Initialiszes jit cache for insert and generate.
+
+  Args:
+      generate_engine: A generate engine to be compiled for.
+      generate_params: The associated parameters.
+      generate_idx: Which generate engine it is.
+  """
+
+  prefill_buckets = token_utils.DEFAULT_PREFILL_BUCKETS
+  prefill_buckets = [
+      bucket
+      for bucket in prefill_buckets
+      if bucket <= generate_engine.max_prefill_length
+  ]
+  generate_engine.prefill_buckets = prefill_buckets
+  if generate_engine.max_prefill_length not in prefill_buckets:
+    prefill_buckets.append(generate_engine.max_prefill_length)
+
+  decode_state = generate_engine.init_decode_state()
```

Comment:
this line will allocate empty kv caches
Then engine will allocate kv caches again.

To compile, you don't need the real input. You can give model jax.ShapedDtypeArray instead. Only shape and dtype need to be specified.